cpp/src/arrow/compute/exec/options.h

```diff
@@ -397,24 +397,51 @@ class ARROW_EXPORT HashJoinNodeOptions : public ExecNodeOptions {
 /// This node will output one row for each row in the left table.
 class ARROW_EXPORT AsofJoinNodeOptions : public ExecNodeOptions {
  public:
-  AsofJoinNodeOptions(FieldRef on_key, FieldRef by_key, int64_t tolerance)
-      : on_key(std::move(on_key)), by_key(std::move(by_key)), tolerance(tolerance) {}
+  AsofJoinNodeOptions(FieldRef on_key, const FieldRef& by_key, int64_t tolerance,
+                      bool nullable_by_key = false)
+      : on_key(std::move(on_key)),
+        by_key(),
+        tolerance(tolerance),
+        nullable_by_key(nullable_by_key) {
+    this->by_key.push_back(std::move(by_key));
+  }
 
-  /// \brief "on" key for the join. Each
+  AsofJoinNodeOptions(FieldRef on_key, std::vector<FieldRef> by_key, int64_t tolerance,
+                      bool nullable_by_key = false)
+      : on_key(std::move(on_key)),
+        by_key(by_key),
```

Comment:
Why do we move `on_key` but not `by_key` here?

Reply:
AFAIK, `std::vector` supports this via copy ellision whereas `FieldRef` supports move-construction.

Reply:
According to [this doc](https://en.cppreference.com/w/cpp/language/copy_elision), copy elision should happen with a move as well.